File: Benchmarks/Core/Systems/ButtonSystem2D.cs
using Benchmarks.Core.Components;
using Benchmarks.Core.Entity;
using Benchmarks.Core.Scene;
using Benchmarks.Core.Systems.Contract;
using Benchmarks.Core.Types;

namespace Benchmarks.Core.Systems
{
    /// <summary>
    /// Sistema encargado de gestionar los botones.
    /// Replica el patron de Rectangle.Contains check del engine.
    /// </summary>
    public class ButtonSystem2D : ISystem
    {
        // Simulated mouse position
        private Vector2 _mousePosition = new(500, 400);
        private bool _mousePressed = false;

        public override void InitializeSystem(EntityManager @entities)
        {
            EntityManager = @entities;
        }

        public override void UpdateSystem(GameTime gameTime)
        {
            if (EntityManager == null) return;

            Span<ButtonComponent2D> buttonComponents = GetButtonComponentsAsSpan();

            foreach (ref Yotsuba entity in GetEntitiesAsSpan())
            {
                if (!entity.HasComponent(YTBComponent.Button2D)) continue;

                ref ButtonComponent2D button = ref buttonComponents[entity.Id];
                if (!button.IsActive) continue;

                if (button.EffectiveArea.Contains((int)_mousePosition.X, (int)_mousePosition.Y))
                {
                    if (_mousePressed)
                    {
                        button.Action?.Invoke();
                    }
                }
            }
        }

        public override void SharedEntityForEachUpdate(ref Yotsuba Entidad, GameTime time) { }
        public override void SharedEntityInitialize(ref Yotsuba Entidad) { }
        public override void Dispose() { }
    }
}

File: Benchmarks_FSharp/Core/Systems/ButtonSystem2D.fs
namespace Benchmarks.Core.Systems

open Benchmarks.Core.Components
open Benchmarks.Core.Entity
open Benchmarks.Core.Scene
open Benchmarks.Core.Systems.Contract
open Benchmarks.Core.Types

type ButtonSystem2D() =
    inherit ISystem()

    let _mousePosition = Vector2(500f, 400f)
    let mutable _mousePressed = false

    override this.InitializeSystem(entities: EntityManager) =
        this.EntityManager <- entities

    override this.UpdateSystem(gameTime: GameTime) =
        if isNull this.EntityManager then ()
        else
        let buttonComponents = this.GetButtonComponentsAsSpan()
        let entities = this.GetEntitiesAsSpan()
        for i = 0 to entities.Length - 1 do
            let entity = &entities.[i]
            if entity.HasComponent(YTBComponent.Button2D) then
                let button = &buttonComponents.[entity.Id]
                if button.IsActive then
                    if button.EffectiveArea.Contains(int _mousePosition.X, int _mousePosition.Y) then
                        if _mousePressed then
                            if not (isNull button.Action) then
                                button.Action.Invoke()

    override _.SharedEntityForEachUpdate(_, _) = ()
    override _.SharedEntityInitialize(_) = ()
    override _.Dispose() = ()
